bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -0,0 +1,440 @@
+   * {@link Scan}.
+   */
+  public static class Source extends BoundedSource<Result> {
+    private static final long serialVersionUID = -5580115943635114126L;
+
+    private final CloudBigtableScanConfiguration configuration;
+
+    /**
+     * A {@link BoundedSource} for a Bigtable {@link Table} with a start/stop key range, along with
+     * a potential filter via a {@link Scan}.
+     */
+    private static class SourceWithKeys extends BoundedSource<Result> {
+      private static final long serialVersionUID = 2561066007121040429L;
+
+      private final CloudBigtableScanConfiguration configuration;
+      private final byte[] startKey;
+      private final byte[] stopKey;
+
+      public SourceWithKeys(
+          CloudBigtableScanConfiguration configuration, byte[] startKey, byte[] stopKey) {
+        this.configuration = configuration;
+        this.startKey = startKey;
+        this.stopKey = stopKey;
+      }
+
+      @Override
+      public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+        return 0;
+      }
+
+      @Override
+      public boolean producesSortedKeys(PipelineOptions options) throws Exception {
+        return true;
+      }
+
+      @Override
+      public Coder<Result> getDefaultOutputCoder() {
+        return new HBaseResultCoder();
+      }
+
+      @Override
+      public List<? extends BoundedSource<Result>> splitIntoBundles(long arg0, PipelineOptions arg1)
+          throws Exception {
+        return Collections.singletonList(this);
+      }
+
+      @Override
+      public void validate() {
+        checkNotNullOrEmpty(configuration.getProjectId(), "project");
+        checkNotNullOrEmpty(configuration.getZone(), "zone");
+        checkNotNullOrEmpty(configuration.getCluster(), "cluster");
+        checkNotNullOrEmpty(configuration.getTable(), "table");
+      }
+
+      @Override
+      public com.google.cloud.dataflow.sdk.io.BoundedSource.BoundedReader<Result> createReader(
+          PipelineOptions options) throws IOException {
+        return new CloudBigtableIO.Source.Reader(this, configuration, startKey, stopKey);
+      }
+    }
+
+    /**
+     * Reads rows for a specific {@link Table}, potentially filted by a Scan and/or a start/stop key
+     * range.
+     */
+    private static class Reader extends AbstractBoundedReader<Result> {
+      private final BoundedSource<Result> source;
+      private final Scan scan;
+      private final CloudBigtableScanConfiguration config;
+
+      private Connection connection;
+      private ResultScanner scanner;
+      private Table table;
+
+      private Result current;
+
+      public Reader(BoundedSource<Result> source, CloudBigtableScanConfiguration config,
+          byte[] startRow, byte[] endRow) throws IOException {
+        super();
+        this.source = source;
+        this.config = config;
+        if (startRow != null || endRow != null) {
+          this.scan = new Scan(config.getScan());
```

Comment:
Do any operations in the Reader constructor perform IO operations ? If so please move them to the start() method.